core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2333,6 +2333,75 @@ private[zio] trait ZIOFunctions extends Serializable {
   final def lock[R, E, A](executor: Executor)(zio: ZIO[R, E, A]): ZIO[R, E, A] =
     new ZIO.Lock(executor, zio)
 
+  /**
+   * Sequentially zips the specified effects using the specified combiner
+   * function.
+   */
+  final def mapN[R, E, A, B, C](zio1: ZIO[R, E, A], zio2: ZIO[R, E, B])(f: (A, B) => C): ZIO[R, E, C] =
```

Comment:
@adamgfraser 
Hmm, I thought the idea was to move the implicit classes on tuples to the companion object? That way they will be picked up without a wildcard import, because conversions are searched also for type parameters and ZIO is present as a type parameter of tuple/Iterable.

Reply:
@neko-kai Sorry I misunderstood you. We could definitely do that, or potentially both variants.

My thinking was that the problem we're trying to solve here is providing a way for users to combine multiple effects. I thought this was better provided through a concrete method on the companion object that doesn't require any implicits. If I have:

```scala
val zio1: ZIO[R, E, A] = ???
val zio2: ZIO[R, E, B] = ???
val zio3: ZIO[R, E, C] = ???
val zio4: ZIO[R, E, D] = ???
val f: (A, B, C, D) => E = ???
```

It seems slightly more natural to do:

```scala
ZIO.mapN(zio1, zio2, zio3, zio4)(f)
```

than:

```scala
(zio1, zio2, zio3, zio4).mapN(f)
```

Both of them are concise and read well but the first is probably more easily discoverable just being a normal method on an object instead of implicit syntax.

Where the new signature is not as convenient is where you already have a tuple of effects, because then you need to destructure the tuple to feed to `mapN`, but I'm not sure how common that is.

I'm happy to do either but that why I thought this was the better solution.

Reply:
@adamgfraser 
I'm fine either way (or both!), just checking if that's intentional. Note though, that plain methods are not equivalent in cases with chaining or variables e.g. ```io.flatMap(_.mapN { ... }))``` – Scala has auto-tupling, but no auto-untupling and I believe `ZIO.mapN.tupled(_) { ... }` will probably not work due to direction of type inference.

Reply:
I like `ZIO.mapN` more as it's obvious where to find it. It's not obvious we add methods to tuples. I've never really liked that style.

I do think `io.flatMap(_.mapN(...))` working only with the syntax is a good point, though. Although in practice it only occurs where an effect produces a tuple. The most obvious candidates for tuple production would be `zip` and `zipPar`, but if we added `ZIO.zipN`, and `ZIO.zipParN`, then suddenly tuple creation doesn't become common anymore, which lessens the argument for it.

🤷‍♂